### PR TITLE
[WIP] Add toggle to not start the OpenShift cluster

### PR DESCRIPTION
**Pull request by gbraad**
Signed-off-by: Gerard Braad <me@gbraad.nl>

This adds a toggle to prevent starting the kubelet service for use with a podman-only target
However, some side-effects remains:

  * a running VM will report `KubeletStarted` (restart)
  * etc.

For a podman-only VM, you can start with `crc start --start-openshift=false`

### CODE REVIEW

File: cmd/crc/cmd/config/config.go [hunk 1]
```diff
@@ -49,6 +50,9 @@ func RegisterSettings(cfg *config.Config) {
 
 	// Telemeter Configuration
 	cfg.AddSetting(ConsentTelemetry, "", config.ValidateYesNo, config.SuccessfullyApplied)
+
+	// Change startup behaviour
+	cfg.AddSetting(StartOpenShift, true, config.ValidateYesNo, config.SuccessfullyApplied)
```
Comment (cfergeau) [inline cmd/crc/cmd/config/config.go:55]: Fwiw, it's not really intuitive to disable seemingly unrelated stuff ('openshift') in order to enable 'podman mode', maybe this should be `cluster-mode: podman|openshift` or something similar.

Comment (gbraad) [reply cmd/crc/cmd/config/config.go:55]: Podman is not a mode; it is always existent.

Comment (guillaumerose) [reply cmd/crc/cmd/config/config.go:55]: Would it be possible to compile crc under a new name like podman-machine and change slightly the behavior based on this ?
If I am using only podman, I use podman-machine but I can "upgrade" instantly by using "crc".

Comment (cfergeau) [reply cmd/crc/cmd/config/config.go:55]: On linux/mac, no need for a new binary, this can be a symlink + a check of os.Args[0]. Maybe this can also work on windows as well.

Comment (cfergeau) [reply cmd/crc/cmd/config/config.go:55]: > Podman is not a mode; it is always existent

Could be `podman-only` and `podman+openshift`. But ok, this `disable-openshift` toggle is not required for podman, it's just an optimization.

Comment (gbraad) [reply cmd/crc/cmd/config/config.go:55]: This should be reviewed in isolation of podman
The start and stop of openshift is something that enables this, but it's not the only driver. 

Starting the vm without anything besides DHCP already is a podman vm with our image

Comment (gbraad) [reply cmd/crc/cmd/config/config.go:55]: The idea could be to reintroduce `crc openshift [start/stop/status/...]`

Comment (guillaumerose) [reply cmd/crc/cmd/config/config.go:55]: > Complexity of a new executable for what benefit? 

Why podman would be a second-class citizen and not openshift ? 
On Linux, users will use crc for openshift. 
On macOS, there is more value and more users to gain with podman.

For me the UX feels better with 2 commands. 

crc is a way to have a maintained docker-machine for podman nothing else.. With 2 separated binaries/or a clean way to differentiate both, I would bake a minimal bundle and create a Podman Desktop.pkg :)

Comment (gbraad) [reply cmd/crc/cmd/config/config.go:55]: you would want to run two VMs too?

Comment (guillaumerose) [reply cmd/crc/cmd/config/config.go:55]: In case of 2 differents apps, yes. Maybe for OKD also, we could be able to launch side by side 2 VMs.

Comment (deboer-tim) [reply cmd/crc/cmd/config/config.go:55]: For me CRC is the 'brand' for single-user/evaluator OpenShift, so even if we agree on a parameter name to have any sort of 'crc --no-openshift' is entirely confusing. The same way that 'java --no-java' or 'chectl --no-che' would be.

It seems like the use of this would be fairly different too, so even if the implementation is shared or very similar I'd prefer to see a separate cli.

Comment (guillaumerose) [reply cmd/crc/cmd/config/config.go:55]: The cost of 2 CLIs is really high.
Maybe we can start with this simple flag and see if people using it with the telemetry. If there is enough usage, we can turn that into a proper CLI ?

Comment (gbraad) [reply cmd/crc/cmd/config/config.go:55]: Right,  it should not be underestimated what the cost involved is. It would be like a totally new lifecycle management tool. Even if shared in this codebase, the technical debt for that is very high.

These changes to have the cluster be restarted bring benefit by itself.

Note: this parameter we also used on minishift for a docker runtime only option.

Comment (gbraad) [reply cmd/crc/cmd/config/config.go:55]: The question can still be; should this be part of CRC?

**Comment by gbraad** [comment]
@deboer-tim this is related to the original scenarios to enable podman to be 're-used' https://github.com/code-ready/crc/issues/1097

the use would be
```
$ crc start 
$ crc podman-env
```

which will start the regular cluster and allows you to use podman alongside the cluster (shares resource usage)

or with this option:
```
$ crc start --start-openshift=false
$ crc podman-env
```

you would have a podman-only VM, but allowing to start the cluster if we add the lifecycle commands like:

```
$ crc openshift start
```

which will start the cluster on the VM if it wasn't running.

**Comment by deboer-tim** [comment]
Thanks for taking the time to walk me through the use cases and background today. Is the ability to change the running CRC a core use-case, or just something that's technically possible? If it hasn't been requested then IMHO it would probably be better to leave out for now and users can just stop and restart.

I agree it's not worth creating a separate download or command line, but it's still bothering me that the cli focuses on 'not starting OpenShift' more than 'We start two things by default but you can choose to only start one of them'. Maybe instead of:
> $ crc start --start-openshift=false
it should be something like:
> $ crc start --podman
or:
> $ crc start --podman-only
?

**Comment by gbraad** [comment]
> it should be something like:
> `$ crc start --podman`
> or:
> `$ crc start --podman-only`
> ?

I would like to leave out referencing `podman` in the flags we have, as they can not easily be removed ince they has once been added. we do not know if the podman support stays

**Comment by cfergeau** [comment]
> > it should be something like:
> > `$ crc start --podman`
> > or:
> > `$ crc start --podman-only`
> > ?
> 
> I would like to leave out referencing `podman` in the flags we have, as they can not easily be removed ince they has once been added. we do not know if the podman support stays (as it has disappeared before :-)

But if it needs to go away, we can rename them while keeping the old option name as an alias, hide the old name, output warnings when the old name is used, before finally removing it. Preferrable not to do that, but does not seem very hard to achieve

**Comment by guillaumerose** [comment]
Proposition: have different levels of crc.

4 levels:
* podman only
* slim openshift: routes and image registry only (recommended for developers - 6GB)
* openshift: all the above + samples/marketplace/olm and console (recommended for trial - 9GB)
* full openshift: all the above + monitoring - (best openshift experience - 12GB)

I imagine at `crc start` time we could prompt the question to the user. `What level of integration do you choose?`
In the tray, we could think at a slider from bottom to top with the 4 levels + description.

The major argument is that with slim openshift people can run it almost all the time.

WDYT?  @gbraad @deboer-tim

**Comment by gbraad** [comment]
I believe we decided to leave CRC untouched and merely expose the `podman-env` command when possible.
We are here still waiting on a usable client as the 2.0.x release did not provide the v2 API (only v2.1).

**Comment by guillaumerose** [comment]
Yes sure, I am just trying to find new ideas/solutions for the future :)

**Comment by deboer-tim** [comment]
I agree it would be nice to have levels or templates, primarily since that would allow us to provide the full version and remove 'missing' capability, while still offering a developer option that's lighter than today.

Personally I'd put the (dev) console in developer, but we could close on details later. :)

**Comment by guillaumerose** [comment]
I made an attempt in the tray. It fits better than a flag like `--only-podman` or `--enable-cluster-monitoring`. 

Screenshot 2021-03-11 at 21 12 51
Screenshot 2021-03-11 at 21 14 03

**Comment by gbraad** [comment]
But can we move the discussion out of the PR and on to the issue perhaps?

**Comment by gbraad** [comment]
Let's continue the discussion here: https://github.com/code-ready/crc/issues/1611 as we have discussed the saeme scenarios before in: https://github.com/code-ready/crc/issues/1097 with agreement of Scott. I believe we need to handle the start and detect the state of the cluster in a better fashion before continuing work. This is why I put this PR on HOLD (and closed now).
